## Chi-squared 

recreating data on this site and then analysing it



link - https://statistics.laerd.com/spss-tutorials/chi-square-test-for-association-using-spss-statistics.php



# Video 1

Make the data / simulate real data from the laerd website

![alternative text](Images/Laerd.Table.jpg)


In [45]:
# data frames

import pandas as pd


# shuffles

import random


# Statistics

import scipy.stats as ss


In [36]:
# sixteen males who preferred books

male_books = [['Male' , 'Books']] * 16


male_books



[['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books']]

In [37]:
# thirteen females who preferred books

female_books = [['Female' , 'Books']] * 13


female_books

[['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books']]

In [38]:
# twenty-four males who preferred online

male_online = [['Male' , 'Online']] * 24


male_online

[['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online']]

In [39]:
# twenty-seven females who preferred online

female_online = [['Female' , 'Online']] * 27


female_online

[['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online']]

Now we have four variables / raw data / lists

we can merge these to then analyse


In [40]:
raw_data = male_books + male_online + female_books + female_online

# this gives us one long list - this is 2 diamentaional list

# the plus joins the end of the list

raw_data

[['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],


In [41]:
## shuffles a list in the outer list not the inner so the values stay the same
# twenty-seven females still like online


random.shuffle(raw_data)

raw_data

[['Male', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Books'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Books'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Ma

In [42]:
# zip the list

# the  raw list contains 80 things and each of those things includes 2 things - male and online for example
# we want to use Pandas
# pandas would prefer - the list to 2 things on the outside and within each of those 80 things. So male and female and within this the options
# this is therefore a quick way to reconfigure the data in a way that pandas will prefer to work with



list(zip(*raw_data))

# unzips a list and puts the inner as outer and the outer as the inner
# rows in colums and columns into rows - interchanges the inner and outer lists
# 2D list *** not sure for a 3d list


# it is a tuple returned and needs to be wrapped in a list for it to be read and nmeeds the *

[('Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female'),
 ('Books',
  'Online',
  'Online',
  'Online',
  'Online',
  'Online',
  'Online',
  'Online',
  'Books',
  'Online',
 

In [43]:
# turn into a dict
# 

gender, medium = list(zip(*raw_data))


# show

gender, medium

# note the differnces in the brackets

(('Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female'),
 ('Books',
  'Online',
  'Online',
  'Online',
  'Online',
  'Online',
  'Online',
  'Online',
  'Books',
  'Online',
 

In [44]:
# create a dataframe

# dataframe - pandasDataframe = dict - name a varible and give it a list of values & name a variable and give it a list of values


df = pd.DataFrame({'gender' : gender, 'medium' : medium})

df

gender  medium
0     Male   Books
1   Female  Online
2   Female  Online
3     Male  Online
4     Male  Online
..     ...     ...
75    Male  Online
76    Male  Online
77    Male   Books
78  Female  Online
79  Female  Online

[80 rows x 2 columns]

So far we have sythasised a dataframe

Set up the variables of male / female - books / online

merged them - long list - long list containing 2 strings

unzipped - so male - then female and then then individual onlien or books




we will now randomise so it looks list " real " data - - - > do before zipping

# Video 2 

Contingency Table

Testing to confirm we have a correct table

In [46]:
# you can call contingency from within the scipy.stats module without having imported it directly in jupyter

ss.contingency.crosstab(df)



CrosstabResult(elements=(array(['Books', 'Female', 'Male', 'Online'], dtype=object),), count=array([29, 40, 40, 51]))

In [48]:
cross = ss.contingency.crosstab(df)

cross 

CrosstabResult(elements=(array(['Books', 'Female', 'Male', 'Online'], dtype=object),), count=array([29, 40, 40, 51]))

In [49]:
cross.elements

(array(['Books', 'Female', 'Male', 'Online'], dtype=object),)

In [50]:
cross.count

array([29, 40, 40, 51])

The above has just treated the df as one long list and counted them.

We want what the intersection is

you can see in the below snip from the scipy site that they have passed in parameters to get the table. Crosstab (a, x)


![alternative text](Images/scipy.crosstab.jpg)

In [51]:
# perform crosstab contingencu

crossA = ss.contingency.crosstab(df['gender'], df['medium'])

crossA

CrosstabResult(elements=(array(['Female', 'Male'], dtype=object), array(['Books', 'Online'], dtype=object)), count=array([[13, 27],
       [16, 24]]))

In [52]:
crossA.elements

(array(['Female', 'Male'], dtype=object),
 array(['Books', 'Online'], dtype=object))

In [53]:
crossA.count

# check this against the table
# we seem to have created the correct table

# female and male along the y axis / books and online on the x axis / female books 13 etc

# we do not as yet have totals etc

array([[13, 27],
       [16, 24]])

In [54]:
# the first variable values, and the second

first , second = crossA.elements

# show
first, second

(array(['Female', 'Male'], dtype=object),
 array(['Books', 'Online'], dtype=object))

In [56]:
df[df['gender'] == first[0]]

# find all rows with gender = to the first value ( female in this case )

gender  medium
1   Female  Online
2   Female  Online
7   Female  Online
8   Female   Books
9   Female  Online
13  Female   Books
14  Female  Online
15  Female  Online
18  Female  Online
19  Female   Books
23  Female  Online
24  Female   Books
28  Female  Online
31  Female  Online
32  Female  Online
33  Female   Books
36  Female   Books
37  Female  Online
39  Female  Online
43  Female  Online
44  Female  Online
45  Female  Online
47  Female   Books
48  Female   Books
49  Female   Books
50  Female  Online
51  Female   Books
53  Female   Books
57  Female  Online
58  Female  Online
59  Female  Online
60  Female  Online
63  Female  Online
66  Female  Online
68  Female   Books
69  Female   Books
73  Female  Online
74  Female  Online
78  Female  Online
79  Female  Online

In [57]:
df[df['gender'] == first[0]][df['medium'] == second[0]]

# find all rows with gender = to the first value ( female in this case ) + medium set to the first value



C:\Users\V\AppData\Local\Temp\ipykernel_1428\3592433814.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['gender'] == first[0]][df['medium'] == second[0]]


gender medium
8   Female  Books
13  Female  Books
19  Female  Books
24  Female  Books
33  Female  Books
36  Female  Books
47  Female  Books
48  Female  Books
49  Female  Books
51  Female  Books
53  Female  Books
68  Female  Books
69  Female  Books

# Video 3 

Do the Statistics

Running the code on the table   

In [60]:
# the contingency table

crossA.count

array([[13, 27],
       [16, 24]])

In [64]:
# statistics from table

result = ss.chi2_contingency(crossA.count, correction=False)

result

Chi2ContingencyResult(statistic=0.486815415821501, pvalue=0.4853513240525321, dof=1, expected_freq=array([[14.5, 25.5],
       [14.5, 25.5]]))

Chi2ContingencyResult(

statistic=0.21636240703177823, 

pvalue=0.6418256702738976, 

dof=1, 

expected_freq=array([[14.5, 25.5],
                     [14.5, 25.5]]))

![alternative text](Images/Laerd.Results.Chi.jpg)


The results here show Pearson Chi is .487 - this does not match out outputs

However their continuity correction does match our output .216

we have added correction=False  // this corrects this issue



![alternative text](Images/Chi.python.correction.jpg)


link > https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html#scipy.stats.chi2_contingency


now our output matches laerds

# Video 4 

Interpret Results

What do the outputs mean?  

Pvalue - this is the key metric for the Chi Square

Where it is small ( or less then 0.5 generally) it suggests you should reject the null hypotesis.

The null hypotesis in this case is that there is no relation between gender and instruction medium.


approx 50% (4.86) chance of seeing results like we did in the table.


In [65]:
# expected frequencies if independent
# 40 females and 40 males

result.expected_freq

array([[14.5, 25.5],
       [14.5, 25.5]])

In [66]:
# break this down - where do these numbers come from?
# there is the same number of female and male participants in this table

# prefered books irrespective of gender
# total books devided by total in table

29 / 80

# result 36.3%

# you will see this listed in the bottom of the table

0.3625

In [67]:
# if no relationship between gender and medium, then we should have same proportion of males and females liking books


40 * (29 / 80 )

# or 40 males by 0.3625
# this matches our expected results

14.5

In [69]:
# 51 prefer online - 40 out of 80 are female

51 * ( 40 / 80 )

# of the 51 the number we expect to be female

25.5

From the above data set we would be close to rejecting the null hypotesis